x86_64-unknown-linux-gnux32 no longer works with 0.4.5

`x86_64-unknown-linux-gnux32` compiled fine with 0.4.4, but no longer does 0.4.5.

I don't actually use that target, but it happened to be in our CI, so i'm letting you know.

```rs
error[E0308]: arguments to this function are incorrect
  --> src/biguint/addition.rs:52:17
   |
52 |         carry = adc(carry, *a, *b, a);
   |                 ^^^        --  --  - expected `&mut u64`, found `&mut u32`
   |                            |   |
   |                            |   expected `u64`, found `u32`
   |                            expected `u64`, found `u32`
   |
   = note: expected mutable reference `&mut u64`
              found mutable reference `&mut u32`
note: function defined here
  --> src/biguint/addition.rs:13:4
   |
13 | fn adc(carry: u8, a: u64, b: u64, out: &mut u64) -> u8 {
   |    ^^^ ---------  ------  ------  -------------
help: you can convert a `u32` to a `u64`
   |
52 |         carry = adc(carry, (*a).into(), *b, a);
   |                            +  ++++++++
help: you can convert a `u32` to a `u64`
   |
52 |         carry = adc(carry, *a, (*b).into(), a);
   |                                +  ++++++++

error[E0308]: arguments to this function are incorrect
  --> src/biguint/addition.rs:57:21
   |
57 |             carry = adc(carry, *a, 0, a);
   |                     ^^^        --     - expected `&mut u64`, found `&mut u32`
   |                                |
   |                                expected `u64`, found `u32`
   |
   = note: expected mutable reference `&mut u64`
              found mutable reference `&mut u32`
note: function defined here
  --> src/biguint/addition.rs:13:4
   |
13 | fn adc(carry: u8, a: u64, b: u64, out: &mut u64) -> u8 {
   |    ^^^ ---------  ------  ------  -------------
help: you can convert a `u32` to a `u64`
   |
57 |             carry = adc(carry, (*a).into(), 0, a);
   |                                +  ++++++++

warning: formatting may not be suitable for sub-register argument
  --> src/biguint/division.rs:48:18
   |
48 |             "div {}",
   |                  ^^
49 |             in(reg) divisor,
   |                     ------- for this argument
   |
   = help: use `{0:e}` to have the register formatted as `eax` (for 32-bit values)
   = help: or use `{0:r}` to keep the default formatting of `rax` (for 64-bit values)
   = note: `#[warn(asm_sub_register)]` on by default

error[E0308]: arguments to this function are incorrect
  --> src/biguint/subtraction.rs:48:18
   |
48 |         borrow = sbb(borrow, *a, *b, a);
   |                  ^^^         --  --  - expected `&mut u64`, found `&mut u32`
   |                              |   |
   |                              |   expected `u64`, found `u32`
   |                              expected `u64`, found `u32`
   |
   = note: expected mutable reference `&mut u64`
              found mutable reference `&mut u32`
note: function defined here
  --> src/biguint/subtraction.rs:13:4
   |
13 | fn sbb(borrow: u8, a: u64, b: u64, out: &mut u64) -> u8 {
   |    ^^^ ----------  ------  ------  -------------
help: you can convert a `u32` to a `u64`
   |
48 |         borrow = sbb(borrow, (*a).into(), *b, a);
   |                              +  ++++++++
help: you can convert a `u32` to a `u64`
   |
48 |         borrow = sbb(borrow, *a, (*b).into(), a);
   |                                  +  ++++++++

error[E0308]: arguments to this function are incorrect
  --> src/biguint/subtraction.rs:53:22
   |
53 |             borrow = sbb(borrow, *a, 0, a);
   |                      ^^^         --     - expected `&mut u64`, found `&mut u32`
   |                                  |
   |                                  expected `u64`, found `u32`
   |
   = note: expected mutable reference `&mut u64`
              found mutable reference `&mut u32`
note: function defined here
  --> src/biguint/subtraction.rs:13:4
   |
13 | fn sbb(borrow: u8, a: u64, b: u64, out: &mut u64) -> u8 {
   |    ^^^ ----------  ------  ------  -------------
help: you can convert a `u32` to a `u64`
   |
53 |             borrow = sbb(borrow, (*a).into(), 0, a);
   |                                  +  ++++++++

error[E0308]: arguments to this function are incorrect
  --> src/biguint/subtraction.rs:75:18
   |
75 |         borrow = sbb(borrow, *ai, *bi, bi);
   |                  ^^^         ---  ---  -- expected `&mut u64`, found `&mut u32`
   |                              |    |
   |                              |    expected `u64`, found `u32`
   |                              expected `u64`, found `u32`
   |
   = note: expected mutable reference `&mut u64`
              found mutable reference `&mut u32`
note: function defined here
  --> src/biguint/subtraction.rs:13:4
   |
13 | fn sbb(borrow: u8, a: u64, b: u64, out: &mut u64) -> u8 {
   |    ^^^ ----------  ------  ------  -------------
help: you can convert a `u32` to a `u64`
   |
75 |         borrow = sbb(borrow, (*ai).into(), *bi, bi);
   |                              +   ++++++++
help: you can convert a `u32` to a `u64`
   |
75 |         borrow = sbb(borrow, *ai, (*bi).into(), bi);
   |                                   +   ++++++++

For more information about this error, try `rustc --explain E0308`.
warning: `num-bigint` (lib) generated 1 warning
error: could not compile `num-bigint` (lib) due to 5 previous errors; 1 warning emitted
```

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -42,6 +42,22 @@ jobs:
       - run: cargo build
       - run: ./ci/test_full.sh
 
+  # try building the x32 target -- x86_64 with target_pointer_width="32"
+  # (we can't execute without kernel CONFIG_X86_X32_ABI though)
+  x32:
+    name: Test (x32)
+    runs-on: ubuntu-latest
+    steps:
+      - run: |
+          sudo apt-get update
+          sudo apt-get install gcc-multilib
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          target: x86_64-unknown-linux-gnux32
+      - run: cargo build --target x86_64-unknown-linux-gnux32 --all-features
+      - run: cargo test --no-run --target x86_64-unknown-linux-gnux32 --all-features
+
   # try a target that doesn't have std at all, but does have alloc
   no_std:
     name: No Std
@@ -78,7 +94,7 @@ jobs:
   success:
     name: Success
     runs-on: ubuntu-latest
-    needs: [test, i686, no_std, fmt, doc]
+    needs: [test, i686, x32, no_std, fmt, doc]
     # Github branch protection is exceedingly silly and treats "jobs skipped because a dependency
     # failed" as success. So we have to do some contortions to ensure the job fails if any of its
     # dependencies fails.

--- a/src/biguint/addition.rs
+++ b/src/biguint/addition.rs
@@ -7,22 +7,32 @@ use core::iter::Sum;
 use core::ops::{Add, AddAssign};
 use num_traits::CheckedAdd;
 
-// Add with carry:
 #[cfg(target_arch = "x86_64")]
-#[inline]
-fn adc(carry: u8, a: u64, b: u64, out: &mut u64) -> u8 {
-    // Safety: There are absolutely no safety concerns with calling `_addcarry_u64`.
-    // It's just unsafe for API consistency with other intrinsics.
-    unsafe { core::arch::x86_64::_addcarry_u64(carry, a, b, out) }
-}
+use core::arch::x86_64 as arch;
 
 #[cfg(target_arch = "x86")]
-#[inline]
-fn adc(carry: u8, a: u32, b: u32, out: &mut u32) -> u8 {
-    // Safety: There are absolutely no safety concerns with calling `_addcarry_u32`.
-    // It's just unsafe for API consistency with other intrinsics.
-    unsafe { core::arch::x86::_addcarry_u32(carry, a, b, out) }
-}
+use core::arch::x86 as arch;
+
+// Add with carry:
+#[cfg(target_arch = "x86_64")]
+cfg_64!(
+    #[inline]
+    fn adc(carry: u8, a: u64, b: u64, out: &mut u64) -> u8 {
+        // Safety: There are absolutely no safety concerns with calling `_addcarry_u64`.
+        // It's just unsafe for API consistency with other intrinsics.
+        unsafe { arch::_addcarry_u64(carry, a, b, out) }
+    }
+);
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+cfg_32!(
+    #[inline]
+    fn adc(carry: u8, a: u32, b: u32, out: &mut u32) -> u8 {
+        // Safety: There are absolutely no safety concerns with calling `_addcarry_u32`.
+        // It's just unsafe for API consistency with other intrinsics.
+        unsafe { arch::_addcarry_u32(carry, a, b, out) }
+    }
+);
 
 // fallback for environments where we don't have an addcarry intrinsic
 // (copied from the standard library's `carrying_add`)

--- a/src/biguint/division.rs
+++ b/src/biguint/division.rs
@@ -44,8 +44,21 @@ fn div_wide(hi: BigDigit, lo: BigDigit, divisor: BigDigit) -> (BigDigit, BigDigi
     unsafe {
         let (div, rem);
 
+        cfg_digit!(
+            macro_rules! div {
+                () => {
+                    "div {0:e}"
+                };
+            }
+            macro_rules! div {
+                () => {
+                    "div {0:r}"
+                };
+            }
+        );
+
         core::arch::asm!(
-            "div {}",
+            div!(),
             in(reg) divisor,
             inout("dx") hi => rem,
             inout("ax") lo => div,

--- a/src/biguint/subtraction.rs
+++ b/src/biguint/subtraction.rs
@@ -7,22 +7,32 @@ use core::cmp::Ordering::{Equal, Greater, Less};
 use core::ops::{Sub, SubAssign};
 use num_traits::CheckedSub;
 
-// Subtract with borrow:
 #[cfg(target_arch = "x86_64")]
-#[inline]
-fn sbb(borrow: u8, a: u64, b: u64, out: &mut u64) -> u8 {
-    // Safety: There are absolutely no safety concerns with calling `_subborrow_u64`.
-    // It's just unsafe for API consistency with other intrinsics.
-    unsafe { core::arch::x86_64::_subborrow_u64(borrow, a, b, out) }
-}
+use core::arch::x86_64 as arch;
 
 #[cfg(target_arch = "x86")]
-#[inline]
-fn sbb(borrow: u8, a: u32, b: u32, out: &mut u32) -> u8 {
-    // Safety: There are absolutely no safety concerns with calling `_subborrow_u32`.
-    // It's just unsafe for API consistency with other intrinsics.
-    unsafe { core::arch::x86::_subborrow_u32(borrow, a, b, out) }
-}
+use core::arch::x86 as arch;
+
+// Subtract with borrow:
+#[cfg(target_arch = "x86_64")]
+cfg_64!(
+    #[inline]
+    fn sbb(borrow: u8, a: u64, b: u64, out: &mut u64) -> u8 {
+        // Safety: There are absolutely no safety concerns with calling `_subborrow_u64`.
+        // It's just unsafe for API consistency with other intrinsics.
+        unsafe { arch::_subborrow_u64(borrow, a, b, out) }
+    }
+);
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+cfg_32!(
+    #[inline]
+    fn sbb(borrow: u8, a: u32, b: u32, out: &mut u32) -> u8 {
+        // Safety: There are absolutely no safety concerns with calling `_subborrow_u32`.
+        // It's just unsafe for API consistency with other intrinsics.
+        unsafe { arch::_subborrow_u32(borrow, a, b, out) }
+    }
+);
 
 // fallback for environments where we don't have a subborrow intrinsic
 // (copied from the standard library's `borrowing_sub`)

